numpyro/util.py

```diff
@@ -164,6 +165,70 @@ def _wrapped(fn):
     return _wrapped
 
 
+def progress_bar_factory(num_samples):
+    """Factory that builds a progress bar decorator along
+    with the `set_tqdm_description` and `close_tqdm` functions
+    """
+
+    tqdm_bars = {}
+
+    def _define_tqdm(arg, transform, device):
+        chain = int(str(device)[4:])
+        tqdm_bars[chain] = tqdm.tqdm(range(num_samples))
+        message = f"Running chain {chain}"
+        tqdm_bars[chain].set_description(message, refresh=False,)
+
+    def _update_tqdm(arg, transform, device):
+        chain = int(str(device)[4:])
+        tqdm_bars[chain].update(arg)
+
+    def _close_tqdm(arg, transform, device):
+        chain = int(str(device)[4:])
+        tqdm_bars[chain].close()
```

Comment:
what happens if `num_samples = 1010` and `print_rate=50`? will this close at 99%?

Reply:
Yes, good suggestion! The progress bar actually overshoots rather than undershoots. I fixed this in the latest commit